### ISSUE TITLE
Binary Operators: Strange formatting of case + pattern match

### ISSUE THREAD

**Issue by david-kubecka** [WhatsApp/erlfmt#163]
```
A = {B, C} = case X of true -> f(); _ -> g() end.
```
gets formatted into
```
A =
    {B, C} =
    case X of
        true -> f();
        _ -> g()
    end.
```
I would expect either
```
A =
    {B, C} =
        case X of
            true -> f();
            _ -> g()
        end.
```
or preferably
```
A = {B, C} =
    case X of
        true -> f();
        _ -> g()
    end.
```

**Comment by awalterschulze** [comment]
This is a very interesting case.  I don't think we have considered this before. Thank you for bringing this up.

**Comment by michalmuskala** [comment]
Interestingly the Elixir formatter does exactly the same thing.
```elixir
a = {b, c} = case d do e -> e end
```
gets formatted to
```elixir
a =
  {b, c} =
  case d do
    e -> e
  end
```

**Comment by michalmuskala** [comment]
I'm going to copy over the reasoning for this behaviour in Elixir that José provided, I find it convincing myself.

> In my opinion, it is intentional. What I do in this case is to replace the operators and expressions and ask if it looks natural. For example, I do expect this:
>
> ```
> foo ++
>  bar ++
>  baz
> ```
> I think indenting baz another time would feel unnatural. So I don’t think we should do it for case. There is an argument though that the first two should be on the same line but it is relatively hard to do with right precedence operators.

**Comment by awalterschulze** [comment]
A great explanation.  It sounds like we would have to make a big exception for equal.

I guess I could also write:
```erlang
A = {{B1, B2}, C} = {B, C} =
    case X of
        true -> f();
        _ -> g()
    end.
```

I also expect this to be very tricky.
@michalmuskala What challenges do you expect with right associativity?

**Comment by michalmuskala** [comment]
The right-precedence makes it complex, because of how expressions are grouped and represented in the AST. This should be more visible if we add explicit parenthesis to represent the grouping:

```erl
A = ({{B1, B2}, C} = ({B, C} =
    case X of
        true -> f();
        _ -> g()
    end)).
```

With right-associative operators we lay out this nested expression
```erl
{B, C} =
    case X of
        true -> f();
        _ -> g()
    end
```
before laying out the wrapping expressions.

We could potentially do a rotation on the AST tree in case the meaning of the resulting program would be the same (i.e. in places where `=` being left- or right-associative would produce the same program). But it is somewhat of a tricky operation.

**Comment by awalterschulze** [comment]
@david-kubecka I am thinking maybe this style is not too bad

```
A =
    {B, C} =
    case X of
        true -> f();
        _ -> g()
    end.
```

I think Michal's investigation and Elixir's answer, makes sense to me.

I am a bit biased by investigating another binary operator issue and starting to understand more what Elixir was saying.
I hope it is okay if we close this issue.

**Comment by awalterschulze** [comment]
This is related to https://github.com/WhatsApp/erlfmt/issues/136 and https://github.com/WhatsApp/erlfmt/pull/198

**Comment by awalterschulze** [comment]
This is fixed in #262

**Comment by awalterschulze** [comment]
Unfortunately we need to reopen this issue as part of https://github.com/WhatsApp/erlfmt/pull/265
We found that our current solution does not support next break fits in `=` chains, see the commented out tests.

**Comment by awalterschulze** [comment]
We really tried to make this better, but we couldn't
For details see our formatting decision document https://github.com/WhatsApp/erlfmt/pull/268
tl;dr we considered a lot of cases and tried several approaches, but the best alternative algorithm was not convincingly better in common cases than what we currently have.  We aren't happy, but realize that sometimes we need to settle.

Personally this is the case that convinced me that formatters are not only subjectively hard, but technically even harder.
